src/System.Memory/src/System/MemoryExtensions.cs

```diff
@@ -17,10 +17,227 @@ namespace System
+        /// <param name="startIndex">The zero-based position to begin deleting characters.</param>
+        /// <param name="count">The number of characters to delete.</param>
+        /// <param name="destination">The span to copy the leftover characters into.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when the destination Span is too small to contain the source Span with the characters removed
+        /// (i.e. destination.Length &lt; source.Length - count).
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when either <paramref name="startIndex"/> or <paramref name="count"/> is less than zero or
+        /// when <paramref name="startIndex"/> plus <paramref name="count"/> specify a position outside the source span.
+        /// </exception>
+        public static void Remove(this ReadOnlySpan<char> source, int startIndex, int count, Span<char> destination)
+        {
+            if ((uint)startIndex > (uint)source.Length || (uint)count > (uint)(source.Length - startIndex))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+            if (destination.Length < source.Length - count)
+                ThrowHelper.ThrowArgumentException_DestinationTooShort();
+
+            // Nothing to remove
+            if (count == 0)
+            {
+                source.CopyTo(destination);
+                return;
+            }
+
+            if (source.Length - count != 0)
+            {
+                if (startIndex == 0)
+                {
+                    source.Slice(count).CopyTo(destination);
+                    return;
+                }
+
+                // destination is after source, reverse the order of copying
+                if (source.Overlaps(destination, out int elementOffset) && elementOffset > 0)
+                {
+                    if (startIndex + count != source.Length)
+                        source.Slice(startIndex + count).CopyTo(destination.Slice(startIndex));
+
+                    // The Slice is necessary since startIndex cannot equal source.Length here.
+                    source.Slice(0, startIndex).CopyTo(destination);
+                }
+                else
+                {   // no overlap or destination is before source
+                    // The Slice is necessary since startIndex cannot equal source.Length here.
+                    source.Slice(0, startIndex).CopyTo(destination);
+
+                    if (startIndex + count != source.Length)
+                        source.Slice(startIndex + count).CopyTo(destination.Slice(startIndex));
+                }
+            }
+            else
+            {   // Remove everything
+                if (source.Length < destination.Length)
+                    destination.Slice(0, source.Length).Clear();
+                else
+                    destination.Clear();
```

Comment:
We are not clearing anything beyond source.Length - count. Here, destination.Length < source.Length.

Reply:
I don't understand. We're only here if source.Length-count is 0. So we shouldn't be clearing anything. What about I missing? There should be nothing to copy and we should be done.

Reply:
Look at this test: https://github.com/dotnet/corefx/pull/26598/files#diff-e1048de19afdbf1765ab34e0dfea880cR162

source = "abc"
destination = "ab"
source.Remove(0, source.Length, destination) // Semantically, we want everything removed here
// Should destination be "/0/0" now or should it remain "ab"?!

I agree that the behavior is more complex than it needs to be, essentially due to the fact that we take in two buffers.

Reply:
If the API does anything more than effectively copy the remaining 1 or 2 pieces, it makes little sense. If everything is removed, the destination should remain untouched; I don't know how to reason about anything else.

Reply:
And if there are no remaining pieces? The destination should remain untouched as well? I am fine with that being the semantics of this method.

Reply:
> And if there are no remaining pieces?

Right, then there's nothing to be done.

Reply:
All that said, I have the same questions about usefulness that Jan raised.  For reference, here's the two places I needed a Remove substitute:
https://github.com/dotnet/coreclr/blob/d08b5fd08ea73bebe4acd1af842ef8f212429944/src/mscorlib/shared/System/Globalization/DateTimeParse.cs#L5296
https://github.com/dotnet/coreclr/blob/d08b5fd08ea73bebe4acd1af842ef8f212429944/src/mscorlib/shared/System/Globalization/DateTimeParse.cs#L5320
and what I used:
https://github.com/dotnet/coreclr/blob/d08b5fd08ea73bebe4acd1af842ef8f212429944/src/mscorlib/shared/System/StringSpanHelpers.cs#L106-L128

Reply:
I had looked at this design and it works just fine because it allocates. Maybe that is the only alternative. A non-allocating Remove() has the quirks that are being discussed here, which potentially reduces its usefulness.

Reply:
If we are fine with allocating, we could re-write RemoveTrailingInQuoteSpaces without using Remove as follows (essentially inlining the call). And this ends up being faster since the user has specific info which means certain operations can be skipped.

```C#
internal void RemoveTrailingInQuoteSpaces()
{
    int i = Length - 1;
    if (i <= 1)
    {
        return;
    }
    char ch = Value[i];
    // Check if the last character is a quote.
    if (ch == '\'' || ch == '\"')
    {
        if (Char.IsWhiteSpace(Value[i - 1]))
        {
            i--;
            while (i >= 1 && Char.IsWhiteSpace(Value[i - 1]))
            {
                i--;
            }

            // Value = Value.Remove(i, Value.Length - 1 - i);
            Span<char> result = new char[i + 1];
            result[i] = ch;
            Value.Slice(0, i).CopyTo(result);
            Value = result.AsReadOnlySpan();
        }
    }
}

internal void RemoveLeadingInQuoteSpaces()
{
    if (Length <= 2)
    {
        return;
    }
    int i = 0;
    char ch = Value[i];
    // Check if the last character is a quote.
    if (ch == '\'' || ch == '\"')
    {
        while ((i + 1) < Length && Char.IsWhiteSpace(Value[i + 1]))
        {
            i++;
        }
        if (i != 0)
        {
            //Value = Value.Remove(1, i);
            Span<char> result = new char[Value.Length - i];
            result[0] = ch;
            Value.Slice(i + 1).CopyTo(result.Slice(1));
            Value = result.AsReadOnlySpan();
        }
    }
}
```